elements/castbar.lua

```diff
@@ -83,443 +85,290 @@ A default texture will be applied to the StatusBar and Texture widgets if they d
+		event = 'UNIT_SPELLCAST_CHANNEL_START'
 	end
 
-	endTime = endTime / 1e3
-	startTime = startTime / 1e3
-	local max = endTime - startTime
+	if(not name or (isTradeSkill and element.hideTradeSkills)) then
+		resetAttributes(element)
+		element:Hide()
 
-	element.castID = castID
-	element.duration = GetTime() - startTime
-	element.max = max
+		return
+	end
+
+	endTime = endTime / 1000
+	startTime = startTime / 1000
+
+	element.max = endTime - startTime
+	element.startTime = startTime
 	element.delay = 0
-	element.casting = true
+	element.casting = event == 'UNIT_SPELLCAST_START'
+	element.channeling = event == 'UNIT_SPELLCAST_CHANNEL_START'
 	element.notInterruptible = notInterruptible
 	element.holdTime = 0
+	element.castID = castID
 	element.spellID = spellID
 
-	element:SetMinMaxValues(0, max)
+	if(element.casting) then
+		element.duration = GetTime() - startTime
+	else
+		element.duration = endTime - GetTime()
+	end
+
+	element:SetMinMaxValues(0, element.max)
 	element:SetValue(0)
 
-	if(element.Text) then element.Text:SetText(text) end
-	if(element.Icon) then element.Icon:SetTexture(texture) end
+	if(element.Icon) then element.Icon:SetTexture(texture or [[Interface\ICONS\INV_Misc_QuestionMark]]) end
+	if(element.Shield) then element.Shield:SetShown(notInterruptible) end
+	if(element.Spark) then element.Spark:Show() end
+	if(element.Text) then element.Text:SetText(name) end
 	if(element.Time) then element.Time:SetText() end
 
-	local shield = element.Shield
-	if(shield and notInterruptible) then
-		shield:Show()
-	elseif(shield) then
-		shield:Hide()
-	end
-
-	local sf = element.SafeZone
-	if(sf) then
-		sf:ClearAllPoints()
-		sf:SetPoint(element:GetReverseFill() and 'LEFT' or 'RIGHT')
-		sf:SetPoint('TOP')
-		sf:SetPoint('BOTTOM')
-		updateSafeZone(element)
-	end
-
-	--[[ Callback: Castbar:PostCastStart(unit, name)
-	Called after the element has been updated upon a spell cast start.
+	local safeZone = element.SafeZone
+	if(safeZone) then
+		local isHoriz = element:GetOrientation() == 'HORIZONTAL'
 
-	* self - the Castbar widget
-	* unit - unit for which the update has been triggered (string)
-	* name - name of the spell being cast (string)
-	--]]
-	if(element.PostCastStart) then
-		element:PostCastStart(unit, name)
-	end
-	element:Show()
-end
+		safeZone:ClearAllPoints()
+		safeZone:SetPoint(isHoriz and 'TOP' or 'LEFT')
+		safeZone:SetPoint(isHoriz and 'BOTTOM' or 'RIGHT')
 
-local function UNIT_SPELLCAST_FAILED(self, event, unit, castID)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+		if(element.casting) then
+			safeZone:SetPoint(element:GetReverseFill() and (isHoriz and 'LEFT' or 'BOTTOM') or (isHoriz and 'RIGHT' or 'TOP'))
+		else
+			safeZone:SetPoint(element:GetReverseFill() and (isHoriz and 'RIGHT' or 'TOP') or (isHoriz and 'LEFT' or 'BOTTOM'))
+		end
 
-	local element = self.Castbar
-	if(element.castID ~= castID) then
-		return
-	end
+		local ratio = (select(4, GetNetStats()) / 1000) / element.max
```

Comment:
Avoid using `select`

Reply:
IMHO, it should be fine in this case, bit I'll remove it if you insist T_T